docs/language-reference/04-types.md

```diff
@@ -0,0 +1,323 @@
+
+```hlsl
+struct S
+{
+    float2 f;
+    int3 i;
+}
+```
+
+### Standard Layout
+
+The _standard layout_ for a structure type uses the following algorithm:
+
+* Initialize variables `size` and `alignment` to zero and one, respectively
+* For each field `f` of the structure type:
+  * Update `alignment` to be the maximum of `alignment` and the alignment of `f`
+  * Set `size` to the smallest multiple of `alignment` not less than `size`
+  * Set the offset of field `f` to `size`
+  * Add the size of `f` to `size`
+
+When this algorithm completes, `size` and `alignment` will be the size and alignment of the structure type.
+
+Most target platforms do not use the standard layout directly, but it provides a baseline for defining other layout algorithms.
+Any layout for structure types must guarantee an alignment at least as large as the standard layout.
+
+### C-Style Layout
+
+C-style layout for structure types differs from standard layout by adding an additional final step:
+
+* Set `size` the smallest multiple of `alignment` not less than `size`
+
+This mirrors the layout rules used by typical C/C++ compilers.
+
+### D3D Constant Buffer Layout
+
+D3D constant buffer layout is similar to standard layout with two differences:
+
+* The initial alignment is 16 instead of one
+
+* If a field would have _improper straddle_, where the interval `(fieldOffset, fieldOffset+fieldSize)` (eclusive on both sides) contains any multiple of 16, *and* the field offset is not already a multiple of 16, then the offset of the field is adjusted to the next multiple of 16
+
+Array Types
+-----------
+
+An _array type_ is either a statically-sized or dynamically-sized array type.
+
+A known-size array type is written `T[N]` where `T` is a type and `N` is a specialization-time constant integer.
+This type represents an array of exactly `N` values of type `T`.
+
+An unknown-size array type is written `T[]` where `T` is a type.
+This type represents an array of some fixed, but statically unknown, size.
+
+### Declaration Syntax
+
+For variable and parameter declarations using traditional syntax, a variable of array type may be declared by using the element type `T` as a type specifier (before the variable name) and the `[N]` to specify the element count after the variable name:
+
+```hlsl
+int a[10];
+```
+
+Alternatively, the array type itself may be used as the type specifier:
+
+```hlsl
+int[10] a;
+```
+
+When using the `var` or `let` keyword to declare a variable, the array type must not be split:
+
+```hlsl
+var a : int[10];
+```
+
+> Note: when declaring arrays of arrays (often thought of as "multidimensional arrays") a programmer must be careful about the difference between the two declaration syntaxes.
+> The following two declarations are equivalent:
+>
+> ```hlsl
+> int[3][5] a;
+> int a[5][3];
+> ```
+>
+> In each case, `a` is a five-element array of three-element arrays of `int`s.
```

Comment:
> Presumably the point here is overall there are the same amount of underlying T in both examples?

No, the point is that those two declarations result in variables with the *exact same type*, despite the order of the element counts in `[]` being different.

This is an obscure but known issue with C-style declaration syntax (one that GLSL ran into head-first when it finally enabled "arrays of arrays").

The C-style declaration rules informally require reading a declaration from the inside out, so we have:

```hlsl
// `a` is a five element array ...
    a[5]   ;

// ... of three-element arrays
    a[5][3];

// .. of ints
int a[5][3];
```

Reading things with types requires going in the other direction, building up the type piece by piece:

```hlsl
// ... int
int

// ... a three-element array of ints
int[3]

// ... a five-element array of three-element arrays of ints
int[3][5]

// `a` is a five-element array of three-element arrays of ints
int[3][5] a;
```

If we could arbitrarily change the syntax, then the Right Way to fix this problem is to make it so that array types are defined the other way around, with `[3]int` instead of `int[3]` for a three-element array of `int`s. With the alternative ordering, then we have:

```hlsl
[5][3]int a; // `a` is a five-element array of three-element arrays of int
```

which looks even better using a more modern `let` or `var` declaration:

```hlsl
var a : [5][3]int;
```

Long story short: the current text in the reference represents the reality of the language and possibly the best possible semantics. The reason I called things out in a note is because of just how confusing the "best possible" semantics can be.